evm/p2p/kademlia.py

```diff
@@ -0,0 +1,446 @@
+            bucket = lower if node.id <= splitid else upper
+            bucket.add_node(node)
+        for node in self.replacement_cache:
+            bucket = lower if node.id <= splitid else upper
+            bucket.replacement_cache.append(node)
+        return lower, upper
+
+    def remove_node(self, node):
+        if node not in self.nodes:
+            return
+        self.nodes.remove(node)
+
+    def in_range(self, node):
+        return self.start <= node.id <= self.end
+
+    @property
+    def is_full(self):
+        return len(self) == self.k
+
+    def add_node(self, node):
+        """
+        If the sending node already exists in the recipient’s k-bucket,
+        the recipient moves it to the tail of the list.
+
+        If the node is not already in the appropriate k-bucket and the bucket has fewer than k
+        entries, then the recipient just inserts the new sender at the tail of the list.
+
+        If the  appropriate k-bucket is full, however, then the recipient pings the k-bucket’s
+        least-recently seen node to decide what to do.
+
+        on success: return None
+        on bucket full: return least recently seen Node for eviction check
+        """
+        self.last_updated = time.time()
+        if node in self.nodes:  # already exists
+            self.nodes.remove(node)
+            self.nodes.append(node)
+        elif len(self) < self.k:  # add if fewer than k entries
+            self.nodes.append(node)
+        else:  # bucket is full
+            return self.head
+
+    @property
+    def head(self):
+        """Least recently seen"""
+        return self.nodes[0]
+
+    @property
+    def depth(self):
+        """Depth is the prefix shared by all nodes in the bucket.
+
+        i.e. The number of shared leading bits.
+        """
+        def to_binary(x):  # left padded bit representation
+            b = bin(x)[2:]
+            return '0' * (k_id_size - len(b)) + b
+
+        if len(self.nodes) < 2:
+            return k_id_size
+
+        bits = [to_binary(n.id) for n in self.nodes]
+        for i in range(1, k_id_size + 1):
+            if len(set(b[:i] for b in bits)) != 1:
+                return i - 1
+        # This means we have at least two nodes with the same ID, so raise an AssertionError
+        # because we don't want it to be caught accidentally.
+        raise AssertionError("Unable to calculate depth")
+
+    def __contains__(self, node):
+        return node in self.nodes
+
+    def __len__(self):
+        return len(self.nodes)
+
+
+class RoutingTable():
+
+    def __init__(self, node):
+        self.this_node = node
+        self.buckets = [KBucket(0, k_max_node_id)]
+
+    def split_bucket(self, bucket):
+        a, b = bucket.split()
+        index = self.buckets.index(bucket)
+        self.buckets[index] = a
+        self.buckets.insert(index + 1, b)
+
+    @property
+    def idle_buckets(self):
+        one_hour_ago = time.time() - k_idle_bucket_refresh_interval
```

Comment:
naming nitpick:

The name `one_hour_ago` doesn't sit well with me.  Either the `k_idle_bucket_refresh_interval` value should be renamed to `one_hour_in_seconds` or we should rename this to `idle_cutoff_timestamp`.